controllers/prep_handlers.go

```diff
@@ -173,53 +169,170 @@ func (r *ImageBasedUpgradeReconciler) queryPrecachingStatus(ctx context.Context,
-	if err := os.Chdir("/"); err != nil {
-		return
+	ostreeRepo := filepath.Join(workspace, "ostree")
+	if err = os.Mkdir(common.PathOutsideChroot(ostreeRepo), 0o755); err != nil {
+		return fmt.Errorf("failed to create ostree repo directory: %w", err)
 	}
 
-	// Write the script
-	scriptname := filepath.Join(utils.IBUWorkspacePath, utils.PrepCleanup)
-	scriptcontent, _ := generated.Asset(utils.PrepCleanup)
-	err := os.WriteFile(common.PathOutsideChroot(scriptname), scriptcontent, 0o700)
+	if _, err = r.Executor.Execute(
+		"tar", "xzf", fmt.Sprintf("%s/ostree.tgz", mountpoint), "--selinux", "-C", ostreeRepo,
+	); err != nil {
+		return fmt.Errorf("failed to extract ostree.tgz: %w", err)
+	}
 
+	// example:
+	// seedBootedId: rhcos-ed4ab3244a76c6503a21441da650634b5abd25aba4255ca116782b2b3020519c.1
+	// seedBootedDeployment: ed4ab3244a76c6503a21441da650634b5abd25aba4255ca116782b2b3020519c.1
+	// seedBootedRef: ed4ab3244a76c6503a21441da650634b5abd25aba4255ca116782b2b3020519c
+	seedBootedId, err := prep.GetBootedStaterootIDFromRPMOstreeJson(filepath.Join(common.PathOutsideChroot(mountpoint), "rpm-ostree.json"))
 	if err != nil {
-		r.Log.Error(err, fmt.Sprintf("failed to write handler script: %s", common.PathOutsideChroot(scriptname)))
+		return fmt.Errorf("failed to get booted stateroot id: %w", err)
+	}
+	seedBootedDeployment := strings.Split(seedBootedId, "-")[1]
+	seedBootedRef := strings.Split(seedBootedDeployment, ".")[0]
+
+	version, err := prep.GetVersionFromClusterInfoFile(filepath.Join(common.PathOutsideChroot(mountpoint), common.ClusterInfoFileName))
+	if err != nil {
+		return fmt.Errorf("failed to get version from ClusterInfo: %w", err)
+	}
+
+	if version != ibu.Spec.SeedImageRef.Version {
+		return fmt.Errorf("version specified in seed image (%s) differs from version in spec (%s)",
+			version, ibu.Spec.SeedImageRef.Version)
+	}
+
+	osname := getStaterootName(version)
+
+	if err = r.OstreeClient.PullLocal(ostreeRepo); err != nil {
+		return fmt.Errorf("failed ostree pull-local: %w", err)
+	}
+
+	if err = r.OstreeClient.OSInit(osname); err != nil {
+		return fmt.Errorf("failed ostree admin os-init: %w", err)
+	}
+
+	kargs, err := prep.BuildKernelArgumentsFromMCOFile(filepath.Join(common.PathOutsideChroot(mountpoint), "mco-currentconfig.json"))
+	if err != nil {
+		return fmt.Errorf("failed to build kargs: %w", err)
+	}
+
+	if err = r.OstreeClient.Deploy(osname, seedBootedRef, kargs); err != nil {
+		return fmt.Errorf("failed ostree admin deploy: %w", err)
+	}
+
+	deploymentID, err := r.RPMOstreeClient.GetDeploymentID(osname)
+	if err != nil {
+		return fmt.Errorf("failed to get deploymentID: %w", err)
+	}
+
+	if _, err = r.Executor.Execute(
+		"cp",
+		filepath.Join(mountpoint, fmt.Sprintf("ostree-%s.origin", seedBootedDeployment)),
+		prep.GetDeploymentOriginPath(osname, deploymentID),
+	); err != nil {
+		return fmt.Errorf("failed to restore origin file: %w", err)
+	}
+
+	if _, err = r.Executor.Execute(
```

Comment:
seems like we are running same command just with another path, lets make it function? it can be in ops.